File: src/maelstrom/net/message.clj
(ns maelstrom.net.message
  "Contains operations specifically related to network messages; used by both
  net and net.journal."
  (:require [schema.core :as s]))



(defrecord Message [^long id src dest body])

(defn message
  "Constructs a new Message. If no ID is provided, uses -1."
  ([src dest body]
   (Message. -1 src dest body))
  ([id src dest body]
   (Message. id src dest body)))

(defn validate
  "Checks to make sure a message is well-formed. Returns msg if legal,
  otherwise throws."
  [m]
  (assert (instance? Message m)
          (str "Expected message " (pr-str m) " to be a Message"))
  (assert (:src m) (str "No source for message " (pr-str m)))
  (assert (:dest m) (str "No destination for message " (pr-str m)))
  m)

File: demo/java/src/main/java/maelstrom/Message.java
package maelstrom;

import com.eclipsesource.json.Json;
import com.eclipsesource.json.JsonObject;
import com.eclipsesource.json.JsonValue;

// A message contains a source and destination, and a JsonObject body.
public class Message implements IJson {
  public final String src;
  public final String dest;
  public final JsonObject body;

  public Message(String src, String dest, JsonObject body) {
    this.src = src;
    this.dest = dest;
    this.body = body;
  }

  public Message(String src, String dest, IJson body) {
    this(src, dest, body.toJson().asObject());
  }

  // Parse a JsonObject as a message.
  public Message(JsonValue j) {
    final JsonObject o = j.asObject();
    this.src  = o.getString("src", null);
    this.dest = o.getString("dest", null);
    this.body = o.get("body").asObject();
  }

  public String toString() {
    return "(msg " + src + " " + dest + " " + body + ")";
  }

  @Override
  public JsonValue toJson() {
    return Json.object()
        .add("src", src)
        .add("dest", dest)
        .add("body", body);
  }
}
